### ISSUE TITLE
Decide on and implement an error-handling policy

### ISSUE THREAD

**Issue by akprasad** [ambuda-org/vidyut#105]
Moving this discussion from #103, with some synthesis of comments from @deepestblue and @shreevatsa --

### Context

I think that vidyut-lipi can become a foundational library for the Sanskrit ecosystem with a useful life of multiple decades. I think so primarily because the Rust ecosystem enables a nice "write once, run anywhere" workflow where we can focus on a single high-quality implementation then bind that implementation elsewhere (Python, Flutter, WASM, ...) as needed.

Foundational libraries should focus on the needs of power users, who expect precision and control. Accordingly, vidyut-lipi's error handling strategy should expose and model cases that a power user might care about.

Such cases include, but are not limited to:

- _Lossy transliteration_ that loses information. Example: Bengali _ba_ which can model Devanagari _ba/va_.
- _Mistaken input_ that is confused for something else. Examples: `:` instead of a *visarga*, `s` instead of an *avagraha*.
- _Malformed input_ that violates our spec. Examples: malformed Grantha numbers, input text that is not in a Unicode standard form (e.g. combining marks out of order).
- _Unknown input_ that cannot be transliterated at all. Examples: text in a different scheme, emojis, some punctuation.

### Prior work

The transliterators I know of generally implement a "best effort" strategy and return a single output string. The very best transliterators, like Aksharamukha, do likewise but also expose a variety of scheme-specific options that let users control how transliteration should proceed.

Since prior work doesn't extensively model error conditions, a natural question is: is error handling worthwhile at all, or is it a pedantic distraction?

I think it's worthwhile in some form (e.g. for malformed Grantha numerals), and I imagine that prior transliterators avoid explicit handling both because of time constraints and because they evolved to suit the needs of specific applications. (@virtualvinodh curious on your thoughts here.)

### Approaches

Our prior discussion in #103 surfaced a variety of approaches to error handling, including:

- best-effort
- failing outright on bad input
- annotating spans of the input/output for their error status

As suggested by @shreevatsa on #33, I like having a two-tier approach:

- a looser high-level API that just gets the user *something* to work with;
- a stricter low-level API that gives the user precise control and insight into the output.

### Prior discussion

See the comments in #103, especially the comments by @deepestblue and @shreevatsa.

I'd like to start by documenting the error cases that might appear, which will inform a specific error-handling strategy.

**Comment by shreevatsa** [comment]
I looked over all the kinds of errors related to input encountered/reported from the Sanskrit metres web app, and most of them fall into the categories already mentioned above:

* Unknown characters in input (possibly already supported): ळ, U+0901 DEVANAGARI SIGN CANDRABINDU, ॐ, ISO 15919 ē ō r̥ r̥̄ l̥ l̥̄ ṁ, UPADHMANIYA and Jihvamuliya (treat as visarga?), फ़ (and क़ख़ग़ज़फ़ड़ढ़)
* Devanagari input that happens to have short e/o for long e/o, or `:` for visarga, or `S` for avagraha: common user error
* ZWNJ/ZWSP in the input, Unicode NFC normalization (combining characters), or NFD (to be able to handle unknown composed characters like `ḿ` at least partially)
* When certain characters in the input are not recognized, warn more prominently.

> I don't believe in the robustness principle

I don't think this is a matter of belief unfortunately :) I remember Mark Pilgrim used to say “Postel's Law has no exceptions” (now can only find [this](http://www.aaronsw.com/weblog/001025) online) — among other things, eventually more tolerant applications will win more users. If vidyut-lipi is intended to be a foundational library for other applications to use, then consider that some of those applications may want to go the "guess what the user intends" route (and offer a "are you sure? / did you mean…?" message or correct automatically). So I think it would be useful for vidyut-lipi to mark those parts of the input (like `:` and `S`) as "suspicious" at least, so that the application can decide whether to silently correct, or warn or educate the user, or fail or whatever.

**Comment by shreevatsa** [comment]
Yeah it wouldn't be good to do any sort of guesswork in the core library itself: as long as an application can achieve this using the library (either preprocessing the input, or postprocessing things like "Unknown" in the output, or passing to the library a list of characters that must be treated specially), it should be fine. Just saying it would be good to design the library so that such usage is possible.
